File: day03/julia/part2.jl
#!/usr/bin/env julia

using FromFile

@from "Helper.jl" import Helper

const OXYGEN = 1
const CO2 = 2

function get_column(i::Int, lines::Vector{String})::Vector{Char}
    return [line[i] for line in lines]
end

function get_value(data::Vector{String}; which=OXYGEN)::Int
    lines = copy(data)

    col_idx = 1
    while length(lines) != 1
        column::Vector{Char} = get_column(col_idx, lines)
        zeros = count(c -> c == '0', column)
        ones = length(column) - zeros
        to_keep::Char = begin
            if which == OXYGEN
                result = '1'    # if ones >= zeros
                if zeros > ones
                    result = '0'
                end
            else    # CO2
                result = '0'    # if zeros <= ones
                if ones < zeros
                    result = '1'
                end
            end
            result
        end
        lines = [line for line in lines if line[col_idx] == to_keep]
        col_idx += 1
    end
    return parse(Int, lines[1], base=2)
end

function main()
    # fname = "example.txt"
    fname = "input.txt"

    data = Helper.read_lines(fname)

    oxygen = get_value(data, which=OXYGEN)
    println(oxygen)
    co2 = get_value(data, which=CO2)
    println(co2)
    println("---")
    println(oxygen * co2)
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: day03/python/part2.py
#!/usr/bin/env python3

import helper

OXYGEN = 1
CO2 = 2


def get_column(i: int, lines: list[str]) -> list[str]:
    return [line[i] for line in lines]


def get_value(data: tuple[str, ...], which=OXYGEN) -> int:
    lines = list(data)

    col_idx = 0
    while len(lines) != 1:
        column: list[str] = get_column(col_idx, lines)
        zeros = column.count('0')
        ones = column.count('1')
        to_keep = ''
        if which == OXYGEN:
            to_keep = '1'    # if ones >= zeros
            if zeros > ones:
                to_keep = '0'
            #
        else:    # CO2
            to_keep = '0'    # if zeros <= ones
            if ones < zeros:
                to_keep = '1'
            #
        #
        lines = [line for line in lines if line[col_idx] == to_keep]
        col_idx += 1
    #
    return int(lines[0], 2)


def main():
    # data = tuple(helper.read_lines("example.txt"))
    data = tuple(helper.read_lines("input.txt"))

    oxygen = get_value(data, which=OXYGEN)
    print(oxygen)
    co2 = get_value(data, which=CO2)
    print(co2)
    print("---")
    print(oxygen * co2)

##############################################################################

if __name__ == "__main__":
    main()
